### PR TITLE
ETCD-180: Bug 1927942: UPSTREAM: <carry>: *: add support for socket options 

### PR DESCRIPTION
**Pull request by hexfusion**
This is a manual backport of https://github.com/etcd-io/etcd/pull/12702 

This PR adds support for setting socket options `SO_REUSEADDR` and `SO_REUSEPORT` to etcd listeners via `ListenConfig`. These options give the flexibility to cluster admins who wish to more explicit control of these features. What we have found is during etcd process restart there can be a considerable time waiting for the port to release as it is held open by `TIME_WAIT` which on many systems is 60s.

```
--socket-reuse-port: enables SO_REUSEPORT [1]  allows rebind of a port already in use. 
--socket-reuse-address: enables SO_REUSEADDR [1]  allows binding to an address in `TIME_WAIT` state. 
```
```
$ ps aux | grep etcd | wc -l
0

$  ss --numeric -o state time-wait 
tcp    0       0                 10.0.138.12:34642            10.0.138.12:2380   timer:(timewait,50sec,0)                                                       
tcp    0       0                 10.0.138.12:34470            10.0.138.12:2380   timer:(timewait,35sec,0)                                                       
tcp    0       0                 10.0.138.12:34086            10.0.138.12:2380   timer:(timewait,10sec,0)                                                       
tcp    0       0                 10.0.138.12:34232            10.0.138.12:2380   timer:(timewait,20sec,0)                                                       
tcp    0       0                 10.0.138.12:34694            10.0.138.12:2380   timer:(timewait,55sec,0)                                                       
tcp    0       0                 10.0.138.12:34332            10.0.138.12:2380   timer:(timewait,25sec,0)                                                       
tcp    0       0                 10.0.138.12:33966            10.0.138.12:2380   timer:(timewait,845ms,0)                                                       
tcp    0       0                 10.0.138.12:34040            10.0.138.12:2380   timer:(timewait,5.845ms,0)                                                     
tcp    0       0                 10.0.138.12:34184            10.0.138.12:2380   timer:(timewait,15sec,0)                                                       
tcp    0       0                 10.0.138.12:34570            10.0.138.12:2380   timer:(timewait,45sec,0)                                                       
tcp    0       0                 10.0.138.12:34512            10.0.138.12:2380   timer:(timewait,40sec,0)                                                       
tcp    0       0                 10.0.138.12:34398            10.0.138.12:2380   timer:(timewait,30sec,0)  
```
So we can wait for many seconds even after etcd process is long dead for client and peer ports to become available.

```
2021-02-19 13:57:25.117257 C | etcdmain: listen tcp 10.0.138.12:2380: bind: address already in use
```

A similar approach has been taken with the addition of these options in the kube-apiserver[2]

[1] https://man7.org/linux/man-pages/man7/socket.7.html
[2] https://github.com/kubernetes/kubernetes/pull/93861

### CODE REVIEW

**Comment by hexfusion** [comment]
etcd appears fine

/retest

**Comment by hexfusion** [comment]
the bug is valid we have issues because we are a fork, manually setting group lead approved as it is not required in 4.8.

File: pkg/transport/timeout_dialer.go [hunk 1]
```diff
@@ -20,17 +20,17 @@ import (
 )
 
 type rwTimeoutDialer struct {
-	wtimeoutd  time.Duration
-	rdtimeoutd time.Duration
+	writeTimeout time.Duration
+	readTimeout  time.Duration
 	net.Dialer
 }
```
Comment (retroflexer) [inline pkg/transport/timeout_dialer.go:26]: The original commit upstream didn't change the names of the fields in this structure.

Comment (hexfusion) [reply pkg/transport/timeout_dialer.go]: yep good catch no sense being greedy.

**Comment by hexfusion** [comment]
/hold 

reviewing extra term churn in aws run.

**Comment by hexfusion** [comment]
CI appears wedged still pending after 7 hours.

/test e2e-aws-serial
/test e2e-aws-upgrade

**Comment by hexfusion** [comment]
CI had outage.. 

/retest

**Comment by hexfusion** [comment]
flake

/retest

**Pull request merged**